lib/semantics/tools.cc

```diff
@@ -794,4 +761,139 @@ static Symbol &InstantiateSymbol(
   return result;
 }
 
+enum class ComponentKind { Direct, Ultimate, Potential };
+
+template<ComponentKind componentKind> class ComponentVisitorImplementation {
+public:
+  ComponentVisitorImplementation(std::function<bool(const Symbol &)> &predicate)
+    : predicate_{predicate} {}
+  SymbolVector Visit(const DerivedTypeSpec &derived) {
+    TraverseDerivedType(derived);
+    return std::move(componentStack_);
+  }
+
+private:
+  const Symbol *TraverseDerivedType(const DerivedTypeSpec &derived) {
+    const Symbol &derivedTypeSymbol{derived.typeSymbol()};
+    // Avoid infinite loop if the type is already part of the types
+    // being visited. It is possible to have "loops in type" because
+    // C744 does not forbid to use not yet declared type for
+    // ALLOCATABLE or POINTER components.
+    // When looking for potential components, the search could fall
+    // in an infinite loop. Avoid these in other search for safety.
+    for (const Symbol *typeUnderVisit : typeStack_) {
```

Comment:
Wouldn't it be better for `typeStack_` to be a set? You wouldn't have to iterate to check membership.

Reply:
Thanks, a stack would have been better for the visitor indeed.
In the iterators, the types used for this check are are kept as `DerivedTypeSpec`, but now the order matter because they are also used to get the scopes, so I could not use a `set` there.